sdk/java/pulumi/src/main/java/com/pulumi/core/Urn.java

```diff
@@ -1,75 +1,312 @@
+            require(t -> isNonEmptyOrNull(t), typeName,
+                    () -> format("expected type, type name part to be not empty, got: '%s'", typeName)
+            );
+            return new Type(package_, module, typeName);
+        }
+
+        public String asString() {
+            return this.package_
+                    + TypeSeparator + this.module.map(m -> m + TypeSeparator).orElse("")
+                    + this.typeName;
+        }
     }
 
     /**
      * Computes a URN from the combination of a resource name, resource type, optional parent,
      * optional project and optional stack.
      */
+    @InternalUse
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
     public static Output<String> create(
-            Output<String> name,
+            Output<String> stack,
+            Output<String> project,
+            Optional<Output<String>> parentUrn,
             Output<String> type,
-            @Nullable Resource parent,
-            @Nullable Output<String> parentUrn,
-            @Nullable Output<String> project,
-            @Nullable Output<String> stack
+            Output<String> name
     ) {
-        Objects.requireNonNull(name);
-        Objects.requireNonNull(type);
+        requireNonNull(name);
+        requireNonNull(type);
 
-        if (parent != null && parentUrn != null) {
-            throw new IllegalArgumentException("Only one of 'parent' and 'parentUrn' can be non-null.");
-        }
-
-        Output<String> parentPrefix;
-        if (parent != null || parentUrn != null) {
-            var parentUrnOutput = parent != null
-                    ? parent.getUrn()
-                    : parentUrn;
-
-            parentPrefix = parentUrnOutput.applyValue(
-                    parentUrnString -> parentUrnString.substring(0, parentUrnString.lastIndexOf("::")) + "$");
-        } else {
-            var stackName = stack == null ? Output.of(Deployment.getInstance().getStackName()) : stack;
-            var projectName = project == null ? Output.of(Deployment.getInstance().getProjectName()) : project;
-            parentPrefix = Output.format("urn:pulumi:%s::%s::", stackName, projectName);
-        }
-
-        return Output.format("%s%s::%s", parentPrefix, type, name);
+        Output<Optional<String>> flippedParentUrn = parentUrn
+                .map(o -> o.applyValue(Optional::of))
+                .orElse(Output.of(Optional.empty()));
+        //noinspection ConstantConditions
+        return Output.tuple(stack, project, flippedParentUrn, type, name)
+                .applyValue(t -> create(t.t1, t.t2, t.t3, t.t4, t.t5));
     }
 
-    public static String create(@Nullable String parent, String type, String name) {
-        Objects.requireNonNull(type);
-        Objects.requireNonNull(name);
+    @InternalUse
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static String create(String stack, String project, Optional<String> parent, String type, String name) {
```

Comment:
Should these not return `URN` instance now instead of `String`?

It is perhaps a breaking change with a large radius but I wanted to discuss it anyway since we're doing this. It would make consuming code a bit clearer, though if the breakage is significant maybe it's not worth it.

Reply:
yes, this is a good idea, I didn't want to do too many changes at once